go-controller/pkg/ovn/egressfirewall.go

```diff
@@ -175,26 +179,56 @@ func (oc *DefaultNetworkController) syncEgressFirewall(egressFirewalls []interfa
 		}
 		existingEFNamespaces[ef.Namespace] = true
 	}
+	// move ACLs from cluster port group to per-namespace port group
+	// old acl matches on the address set, which is still valid. match change form address set to port group will be
+	// performed by the egress firewall handler.
+	// sync the ovn and k8s egressFirewall states.
 	predicateIDs := libovsdbops.NewDbObjectIDs(libovsdbops.ACLEgressFirewall, oc.controllerName, nil)
 	aclP := libovsdbops.GetPredicate[*nbdb.ACL](predicateIDs, nil)
 	efACLs, err := libovsdbops.FindACLsWithPredicate(oc.nbClient, aclP)
 	if err != nil {
 		return fmt.Errorf("cannot find Egress Firewall ACLs: %v", err)
 	}
-	deleteACLs := []*nbdb.ACL{}
-	for _, efACL := range efACLs {
-		if !existingEFNamespaces[efACL.ExternalIDs[libovsdbops.ObjectNameKey.String()]] {
-			// there is no egress firewall in that namespace, delete
-			deleteACLs = append(deleteACLs, efACL)
-		}
-	}
 
-	err = libovsdbops.DeleteACLsFromPortGroups(oc.nbClient, []string{types.ClusterPortGroupNameBase}, deleteACLs...)
-	if err != nil {
-		return err
+	var nsACLs = map[string][]*nbdb.ACL{}
+	for _, acl := range efACLs {
+		namespace := acl.ExternalIDs[libovsdbops.ObjectNameKey.String()]
+		// no key will be treated as an empty namespace and ACLS will just be deleted
+		nsACLs[namespace] = append(nsACLs[namespace], acl)
 	}
-	klog.Infof("Deleted %d stale egress firewall ACLs", len(deleteACLs))
-	return nil
+
+	err = batching.BatchMap[*nbdb.ACL](aclChangePGBatchSize, nsACLs, func(batchNsACLs map[string][]*nbdb.ACL) error {
+		var ops []libovsdb.Operation
+		var err error
+		for namespace, acls := range batchNsACLs {
+			if namespace != "" {
+				pgName := oc.getNamespacePortGroupName(namespace)
+				// delete stale ACLs from namespaced port group
+				if !existingEFNamespaces[namespace] {
+					ops, err = libovsdbops.DeleteACLsFromPortGroupOps(oc.nbClient, ops, pgName, acls...)
+				} else {
+					// re-attach from ClusterPortGroupNameBase to namespaced port group.
+					// port group should exist, because namespace handler will create it.
+					ops, err = libovsdbops.AddACLsToPortGroupOps(oc.nbClient, ops, pgName, acls...)
+				}
+				if err != nil {
+					return fmt.Errorf("failed to build cleanup ops: %w", err)
+				}
+			}
+			// delete all EF ACLs from ClusterPortGroupNameBase
+			ops, err = libovsdbops.DeleteACLsFromPortGroupOps(oc.nbClient, ops,
+				oc.getClusterPortGroupName(types.ClusterPortGroupNameBase), acls...)
+			if err != nil {
+				return fmt.Errorf("failed to build cleanup from ClusterPortGroup ops: %w", err)
+			}
+		}
```

Comment:
Now that we have one PG for the namespace, and if that PG becomes stale because the namespace was deleted, then the PG will be deleted by the namespace sync function and thus the ACLs will be garbage collected with it, we wouldn't need to explicitly delete those ACLs from here. 

You might still be doing this for the UTs, but when talking about potential massive ops I would rather handle it specifically in the UTs than in the real code. In any case it might be appropriate to add a comment.

So overall I am looking at the added startup time here because of unconditional massive ops that we might end up doing. Would it make sense to read the cluster PG, and just move the EF  ACLs that are still attached to it? Ideally once this thing, that is costing us time, has been done once, we wouldn't need to do it again.

Reply:
it's almost right, but the problem is that a namespace may not only be deleted, but also just get an egress firewall deleted, in that case it won't get to `existingEFNamespaces` and a cleanup is needed. That is why the most part will have to stay, but I added an extra condition for clusterPG to reduce the number of ops.

Reply:
Okay! My final nit would be to reorganize the method a bit. It contains several migrations from several different versions and the thing is getting big and confusing. I would probably split them out to their own methods for clarity. But I know I am pushing it now, so up to you!

Reply:
that makes sense :) hope I didn't break anything